File: 221.最大正方形.cpp
/*
 * @lc app=leetcode.cn id=221 lang=cpp
 *
 * [221] 最大正方形
 */

// @lc code=start
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        vector<vector<int>> dp(m+1, vector<int>(n+1, 0));
        int maxArea = 0;
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (matrix[i-1][j-1] == '1') {
                    dp[i][j] = 1 + min(min(dp[i-1][j], dp[i][j-1]), dp[i-1][j-1]);
                    maxArea = max(maxArea, dp[i][j]*dp[i][j]);
                }
            }
        }
        return maxArea;
    }
};
// @lc code=end

File: 221.最大正方形.py
#
# @lc app=leetcode.cn id=221 lang=python3
#
# [221] 最大正方形
#

# @lc code=start
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        m, n = len(matrix), len(matrix[0])
        prefix = [[0]*(n+1) for _ in range(m+1)]
        for i in range(1, m+1):
            for j in range(1, n+1):
                prefix[i][j] = (0 if matrix[i-1][j-1] == '0' else 1) + prefix[i-1][j] + prefix[i][j-1] - prefix[i-1][j-1]
        for l in range(min(m, n), 0, -1):
            for i in range(m-l+1):
                for j in range(n-l+1):
                    if prefix[i+l][j+l] - prefix[i][j+l] - prefix[i+l][j] + prefix[i][j] == l*l:
                        return l*l
        return 0
# @lc code=end

